Gradle 7.1 throws a NullPointerException if launched on JRE on Windows

The build was triggered on a TFS server (Windows)
`C:\Windows\system32\cmd.exe /D /S /C "C:\agent\_work\694\s\gradlew.bat --debug --stacktrace build"`

It turned out that by mistake JRE was used instead of JDK:
```
2021-08-12T20:08:50.7407518Z 2021-08-12T23:08:50.705+0300 [DEBUG] [org.gradle.launcher.daemon.client.DefaultDaemonStarter] Starting daemon process: workingDir = C:\Users\buildagent\.gradle\daemon\7.1.1, daemonArgs: [C:\Program Files\Java\jre1.8.0_301\bin\java.exe, -XX:MaxMetaspaceSize=256m, -XX:+HeapDumpOnOutOfMemoryError, -Xms256m, -Xmx512m, -Dfile.encoding=UTF8, -Duser.country=US, -Duser.language=en, -Duser.variant, -cp, C:\Users\buildagent\.gradle\wrapper\dists\gradle-7.1.1-all\1wqbeia0d49252rmlcr6o8lbl\gradle-7.1.1\lib\gradle-launcher-7.1.1.jar, org.gradle.launcher.daemon.bootstrap.GradleDaemon, 7.1.1]
```

The build failed with just a NullPointerException:
```
2021-08-12T22:01:44.232+0300 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] FAILURE: Build failed with an exception.
2021-08-12T22:01:44.233+0300 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
2021-08-12T22:01:44.235+0300 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * What went wrong:
2021-08-12T22:01:44.235+0300 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] java.lang.NullPointerException
2021-08-12T22:01:44.236+0300 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] > java.lang.NullPointerException (no error message)
2021-08-12T22:01:44.237+0300 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
```
I had to add `--stacktrace` parameter to see what the cause was

### Expected Behavior
Gradle should emit a meaningful message that JDK is required instead of JRE

### Current Behavior
Gradle throws a NullPointerException: 
[stacktrace.txt](https://github.com/gradle/gradle/files/6978349/stacktrace.txt)

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/SimpleGeneratedJavaClassCompiler.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/SimpleGeneratedJavaClassCompiler.java
@@ -44,6 +44,9 @@ class SimpleGeneratedJavaClassCompiler {
      */
     public static void compile(File srcDir, File dstDir, List<ClassSource> classes, ClassPath classPath) throws GeneratedClassCompilationException {
         JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
+        if (compiler == null) {
+            throw new GeneratedClassCompilationException("No Java compiler found, please ensure you are running Gradle with a JDK");
+        }
         DiagnosticCollector<JavaFileObject> ds = new DiagnosticCollector<>();
         try (StandardJavaFileManager mgr = compiler.getStandardFileManager(ds, null, null)) {
             List<String> options = buildOptions(dstDir, classPath);

